arch/armv7a/signal.S

```diff
@@ -20,15 +20,14 @@
 .globl _signal_trampoline
 .type _signal_trampoline, %function
 _signal_trampoline:
-	// stack: return address, signal number
-	push {r0-r3,r9,r12,lr}
-	mrs r0, apsr
-	push {r0}
-	ldr r0, [sp, #32]
-	blx _signal_handler
+	/* Get signal number from stack */
 	pop {r0}
-	msr apsr_nzcvqg, r0
-	pop {r0-r3,r9,r12,lr}
-	add sp, #4
-	pop {pc}
+	blx _signal_handler
+
+	/* Old signal mask in r0 */
+	pop {r1} /* cpu context * */
+	pop {r2} /* pc */
+	pop {r3} /* sp */
```

Comment:
We need to keep stack aligned to 8 at all times i.e. always pop/push multiples of 2 at once